### DISCUSSION TITLE
Automatically seal not-derived internal and private classes

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#4573] [General]
I saw this:

https://github.com/dotnet/runtime/commit/bdc8e420aa75999021e06b85e2e1869962730a0f

And it made me think that, since the compiler knows if a class is derived, this is something the compiler could do automatically. Or at least offer a code fix for it.

**Comment by yaakov-h** [comment]
https://github.com/dotnet/runtime/issues/49944 will offer a code fix for it, assuming the project goes ahead.

**Reply by DavidArno** [reply]
Interesting. I didn't realise there were performance benefits to sealing classes. I seal everything by default as (in my opinion; other opinions are available) it makes for more robust and reliable code as derived classes have to be planned and intended. I can now add the "and it increases the performance of the code" line to my argument arsenal now for seal-by-default. And as a bonus, an "official" analyzer for prompting me to remember to seal even public classes would be amazing. 😃

**Comment by YairHalberstadt** [comment]
As discussed in https://github.com/dotnet/runtime/issues/49944 this can't be done automatically because types can be created at runtime via reflection.Emit

**Reply by AartBluestoke** [reply]
agreed, one my libraries generates types derived from a central type at runtime; i think if things with no children auto-sealed, my code would fail at least 4 different ways.

**Comment by theunrepentantgeek** [comment]
The compiler _can't_ know enough to seal automatically; but the developer can - and it's an easy change.

**Comment by canton7** [comment]
See this discussion on why the JIT can't automatically seal types: https://github.com/dotnet/runtime/issues/49944 . Many of those arguments also apply here.
